### PR TITLE
First kokkos implementation

### PR DESCRIPTION
**Pull request by tpadioleau**

### CODE REVIEW

**Comment by jbigot** [comment]
What about closing this PR and making a dedicated branch?

**Comment by tpadioleau** [comment]
What do you mean ? an other branch ?

**Comment by jbigot** [comment]
the name of this one sounds good to me

**Comment by tpadioleau** [comment]
I use the `draft PR` to show that there is a work in progress on a topic.

**Comment by jbigot** [comment]
ok

File: tests/chunk.cpp [hunk 1]
```diff
@@ -526,7 +526,9 @@ TEST(Chunk2DTest, DeepcopyReordered)
         }
     }
     ChunkYX<double> chunk2(select<DDimY, DDimX>(chunk.domain()));
-    deepcopy(chunk2, chunk);
+    ChunkSpan<double, DDomXY, Kokkos::HostSpace, std::experimental::layout_left>
+            chunk2_view(chunk2.data(), chunk.domain());
+    deepcopy(chunk2_view, chunk);
```
Comment (tpadioleau) [inline tests/chunk.cpp:531]: I suggest we offer a mechanism to reorder the tags and compute the new layout. For example:
```cpp
Chunk<double, DDomXY, MemorySpace, stdex::layout_right> v1;
Chunk<double, DDomYX, MemorySpace, stdex::layout_left> v2(select<Y, X>(v1));
```
It could be with a free function or the slicing operator using a tagged `full_extent`.

File: docs/first_steps/heat_equation.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
+#include <ddc/Chunk>
```
[the remark this answers is not in the document]
Comment (tpadioleau) [reply docs/first_steps/heat_equation.cpp:1]: nope indeed

**Comment by tpadioleau** [comment]
Is breaking voicexx for the moment.

File: benchmarks/deepcopy.cpp [hunk 3]
```diff
@@ -17,7 +17,7 @@ using DVectX = DiscreteVector<DDimX>;
 using DDomX = DiscreteDomain<DDimX>;
 
 template <class Datatype>
-using ChunkSpanX = ChunkSpan<Datatype, DDomX>;
+using ChunkSpanX = ChunkSpan<Datatype, DDomX, Kokkos::HostSpace>;
```
Comment (jbigot) [inline benchmarks/deepcopy.cpp:20]: could this be the default so as not to change everything?

File: examples/game_of_life.cpp [hunk 4]
```diff
@@ -51,6 +57,8 @@ std::ostream& print_2DChunk(
 
 int main()
 {
+    Kokkos::ScopeGuard scope;
```
Comment (jbigot) [inline examples/game_of_life.cpp:60]: can we wrap that?

File: examples/game_of_life.cpp [hunk 5]
```diff
@@ -17,18 +17,24 @@ static unsigned height = 5;
 
 void blinker_init(
         DiscreteDomain<DDimX, DDimY> const& domain,
-        Chunk<cell, DiscreteDomain<DDimX, DDimY>>& cells)
+        ChunkSpan<
+                cell,
+                DiscreteDomain<DDimX, DDimY>,
+                Kokkos::DefaultExecutionSpace::memory_space> cells)
```
Comment (jbigot) [inline examples/game_of_life.cpp:23]: could we wrap it?

File: examples/game_of_life.cpp [hunk 6]
```diff
@@ -81,10 +100,12 @@ int main()
                     for (int i = -1; i < 2; ++i) {
                         for (int j = -1; j < 2; j++) {
                             if (cells_in(ix + i, iy + j)) {
-                                std::cout << "ix : " << ix
-                                          << " | i : " << i
-                                          << " | ix + i : " << ix + i
-                                          << "\n";
+                                // std::printf(
```
Comment (jbigot) [inline examples/game_of_life.cpp:103]: maybe remove it?

File: examples/game_of_life.cpp [hunk 7]
```diff
@@ -101,8 +122,9 @@ int main()
                     }
                 });
         deepcopy(cells_in, cells_out);
+        deepcopy(cells_in_host_alloc, cells_in);
```
Comment (jbigot) [inline examples/game_of_life.cpp:125]: why?

File: examples/heat_equation.cpp [hunk 8]
```diff
@@ -189,33 +195,44 @@ int main()
...
 
+    Chunk<double, DiscreteDomain<DDimX, DDimY>, HostAllocator<double>>
+            ghosted_temp(DiscreteDomain<
+                         DDimX,
+                         DDimY>(ghosted_x_domain, ghosted_y_domain));
+
+
     //! [initial output]
     // display the initial data
+    deepcopy(ghosted_temp, ghosted_last_temp);
```
Comment (jbigot) [inline examples/heat_equation.cpp:233]: is is possible to deepcopy a slice?

File: include/ddc/chunk.hpp [hunk 9]
```diff
@@ -241,6 +248,38 @@ class Chunk<ElementType, DiscreteDomain<DDims...>, Allocator>
         return base_type::allocation_mdspan();
     }
 
+    /** Provide a mdspan on the memory allocation
+     * @return allocation mdspan
+     */
+    constexpr auto allocation_kokkos_view()
```
Comment (jbigot) [inline include/ddc/chunk.hpp:254]: do we need to expose kokkos specific stuff?

File: tests/chunk.cpp [hunk 10]
```diff
@@ -526,7 +526,9 @@ TEST(Chunk2DTest, DeepcopyReordered)
         }
     }
     ChunkYX<double> chunk2(select<DDimY, DDimX>(chunk.domain()));
-    deepcopy(chunk2, chunk);
+    ChunkSpan<double, DDomXY, Kokkos::HostSpace, std::experimental::layout_left>
```
Comment (jbigot) [inline tests/chunk.cpp:529]: why is this change required?

File: tests/non_uniform_discretization.cpp [hunk 11]
```diff
@@ -36,35 +38,35 @@ TEST(NonUniformDiscretizationTest, Rank)
 
 TEST(NonUniformDiscretizationTest, ListConstructor)
 {
-    DDimX ddim_x(VALUES);
+    DDimX::Impl<Kokkos::HostSpace> ddim_x(VALUES);
```
Comment (jbigot) [inline tests/non_uniform_discretization.cpp:41]: isn't this the default?

**Review by jbigot — APPROVED** [review]
